### DISCUSSION TITLE
More Detailed Windows setup

### DISCUSSION THREAD

**Discussion by kpcenti** [nschlia/ffmpegfs#166] [Q&A]
Hello,
I am not sure if this is the correct place to inquire.

I'm trying to follow this guide on the Windows build and the directions can get unclear easily. https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md

For example the "Install WinFSP Support For Cygwin" says to follow the instructions on the winfsp website, and their directions are a little unclear since the installer already has an option for FUSE for Cygwin so i'm not sure if I'm suppose to run those commands.

I'm also not sure how and where to run the commands from the "Install Packages From Cygwin Setup" while it says they should already be available with the setup. 

The Windows branch (https://github.com/nschlia/ffmpegfs/tree/windows) has a lot of references to Linux which clicking on some links leads you back the main branch like Install link under "Installation Instructions" 

I hope this didn't come as silly, I would really like to try this program out.

Thanks,

**Comment by nschlia** [comment]
Probably I was a bit careless writing the guide. I'll review and enhance the documentation. Thanks for pointing this out.

**Reply by kpcenti** [reply]
Thank you! I'm not super versed in advanced setup, but I would really like to use this.

**Comment by nschlia** [comment]
I have edited the document, hope things are clearer now. Please give it a try and let me know.

**Comment by kpcenti** [comment]
The building [Building libbluray](https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md#build-libbluray) link in the https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md leads no where.

**Reply by nschlia** [reply]
> The building [Building libbluray](https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md#build-libbluray) link in the https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md leads no where.

Fixed right away by removing it. Locally, in my editor, it worked. Seems that the Github renderer does not respect that.

**Comment by kpcenti** [comment]
I'm finally following through the tutorial. I went through and installed all the packages one by one.
I'm stuck in the Building Libchardet section. Where do I run those commands in? Cygwin or in a windows terminal? 

I tried cygwin and I got to $ cd install and the rest of the commands do not work it seems

**Reply by nschlia** [reply]
Yes, this must be run inside a Cygwin terminal. I've added a note to make that clear.

But, what is the problem? Any error messages?

I'v followed the instructions and successfully build/installed libchardet:

[build.log](https://github.com/user-attachments/files/17947710/build.log)

**Reply by kpcenti** [reply]
I think the formatting of the text in the commands section there gets jumbled.
Following it line by line ends up getting stuck at line 3
![image](https://github.com/user-attachments/assets/e1bb0673-673a-4e6a-869c-594de3b886dd)

It seems I just have to ignore the ~/install and $ and these lines 

Length: 435028 (425K) [application/octet-stream]
Saving to: 'libchardet-1.0.6.tar.bz2'
2023-01-07 12:19:06 (6.41 MB/s) - 'libchardet-1.0.6.tar.bz2' saved [435028/435028]

Copying those seems to not work and redundant since the previous command makes those lines show up.



I'm not entirely sure how the syntax works for these commands, but it does get confusing  but that might just be me and user error.

Thanks

**Reply by nschlia** [reply]
Yes, the correct command is just "wget ...". "~/install $" is part of the prompt. I added line breaks to the instructions to make that clearer. Of yourse, the $ does not belong to the command, still.

**Comment by kpcenti** [comment]
I think I'm finally close to finally using ffmpegfs. The location to place ffmpeg seems very vague on Windows. I placed mine in C:/win/ffmpeg/win64/ 

Now i'm at building the binaries for FFmpegfs and stuck at the first command of ./configure .

I really do apologize for all these pesky questions.

**Reply by nschlia** [reply]
You're welcome: This way I can get my documentation bulletproof.

The FFmpeg binaries go into win/ffmpeg/win64/ under the FFmpegfs sources directory. When you unpack the tarball, you will find that directory. Alas, the paragraph that talks about that comes right after the FFmpeg instructions. Optimally confusing :)

Sorry 'bout that. I've updated this section: https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md#build-the-binaries-for-ffmpegfs

**Reply by kpcenti** [reply]
Is the regular release of ffmpegfs ok to use for the windows directions, not sure how branches work but the one in the guide leads to the regular branch it seems. 
I download the tar.gz to my C: drive. Got cygwin terminal to cd and unpack using tar. When it mentions to cd to the source, is that the root of the ffmpegfs-2.17 folder or in the src? It seems a little unclear

I managed to run the ./configure command at the root and it seems to work. It did give a worrying error of "Package 'fuse3' not found" which is odd cause that would have been installed with the installer. 
(EDIT: I missed the step running the FUSE commands, this error does not appear anymore but the next one does and updated image)

I continued and tried to run make and was given make: *** No targets specified and no makefile found.  Stop.


![image](https://github.com/user-attachments/assets/2ab25942-cf54-4a6c-a8b6-5ca9c5afcb64)

**Reply by nschlia** [reply]
No, the main branch or the official release packages won't work under Windows. The Debian release has been updated for Fuse 3, but the WinSFP package provides Fuse 2 only. This is why configure failed (Requirements not met).

When you clone the windows branch you should be able to build.

Use "git clone https://github.com/nschlia/ffmpegfs.git -b windows" to get the Windows branch and try the configure/make thing,

**Reply by nschlia** [reply]
Oh, I just noticed the documentation was totally wrong: Of course you cannot download the official release and build it for windows anymore.

I've edited that part: https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md#build-the-binaries-for-ffmpegfs

This what I wrote in my previous post: Clone the Windows branch, and build that.

**Reply by kpcenti** [reply]
So I've downloaded/cloned the windows branch and now the ./configure command doesn't work anymore.
![image](https://github.com/user-attachments/assets/e3bdc1ab-d8a0-4f31-a449-56862486eef0)

**Reply by nschlia** [reply]
Oh no! I forgot to mention you first need to run ./autogen.sh. This will generate the configure script. This is why you needed to install all the the automake tools as prerequisites in the first place.

Sorry again, I've edited the instructions again.

So, run
```
./autogen.sh
./configure
make check
```
That should do it.

**Comment by kpcenti** [comment]
So close!

I finished the last step. I did the make check, which pass-ed most of it, it had some FAILs for some. I did the export, and set env.

I tried running the exported exe in cygwin64 terminal and was met with this
![image](https://github.com/user-attachments/assets/e806f073-358a-4dce-8275-870f1cae0214)


For giggles I tried running the exe in CMD and just double clicking, I get these errors:
![image](https://github.com/user-attachments/assets/46928115-71dd-4891-b574-3d2773616650)

it says I need avcodec-61.dll, avfilter-10.dll, avformat-61.dll, avutil-59.dll

**Reply by nschlia** [reply]
Congrats! You made it 👍 

You cannot run ffmpegfs from a CMD prompt, it's a Cygwin executable, but you successfully did in from Cygwin. To get help, the command would be "ffmpegfs -h". I just noticed that the help output is garbled in the end, I'll fix that, but that won't do any harm so far.

To use ffmpegfs, you need to set some parameters. I'v added a "Usage" paragraph: https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md#usage

Example:
```
mkdir out
ffmpegfs --audiobitrate=256K --videobitrate=1.5M  /cygdrive/c out -o allow_other,ro,desttype=mp4
```
This will map C: to $HOME/out. Just change drives and paths.

PS:

On my system, this fails:

```
$ ffmpegfs --audiobitrate=256K --videobitrate=1.5M  /cygdrive/c out -o allow_other,ro,desttype=mp4
Cannot create WinFsp-FUSE file system: mount point in use.
```

Hope it works for you - I'll have to find out what's wrong here.

**Reply by kpcenti** [reply]
Yep, that was my bad. I read executable and command prompt and thought it ran in windows cmd. Is there a print out of the -h command text as like a text file? towards the end does get garbled and it's hard to read everything in the terminal.

Thank you so much!!

**Reply by kpcenti** [reply]
This might be out of scope for this thread but I wasn't able to run any of the commands and just get command not found for any commands
I'm basically trying to have the output file as ALAC.

![image](https://github.com/user-attachments/assets/1ee1bad6-d20e-4342-939c-6183ad1cf157)

**Reply by nschlia** [reply]
> Yep, that was my bad. I read executable and command prompt and thought it ran in windows cmd. Is there a print out of the -h command text as like a text file? towards the end does get garbled and it's hard to read everything in the terminal.
> 
Yes, I already noticed (see previous post). I've created an issue #167.

**Reply by nschlia** [reply]
> This might be out of scope for this thread but I wasn't able to run any of the commands and just get command not found for any commands I'm basically trying to have the output file as ALAC.
> 
You must run ffmpeg.exe from the same directory as `. setenv`. Do not cd to export.

If you get "Cannot create WinFsp-FUSE file system: mount point in use." like I did, please create an issue. All this worked before, the code has not been changed, I don't know what's wrong all of a sudden.

**Reply by kpcenti** [reply]
> > This might be out of scope for this thread but I wasn't able to run any of the commands and just get command not found for any commands I'm basically trying to have the output file as ALAC.
> 
> You must run ffmpeg.exe from the same directory as `. setenv`. Do not cd to export.
> 
> If you get "Cannot create WinFsp-FUSE file system: mount point in use." like I did, please create an issue. All this worked before, the code has not been changed, I don't know what's wrong all of a sudden.



ah I see! would this be the correct way to call ffmpegfs.exe while staying in the main directory ?

$ /cygdrive/c/ffmpegfs/export/ffmpegfs.exe [options]

![image](https://github.com/user-attachments/assets/b4fdb219-8c1e-4434-8398-976317dd0492)

Unfortunately I also get the mount point in use error. I'll be making a issue momentarily

**Reply by nschlia** [reply]
> ah I see! would this be the correct way to call ffmpegfs.exe while staying in the main directory ?

Simply run ffmpegfs.exe from the build directory, /cygdrive/c/ffmpegfs for you. Like you did here:

![image](https://github.com/user-attachments/assets/e806f073-358a-4dce-8275-870f1cae0214)

> 
> Unfortunately I also get the mount point in use error. I'll be making a issue momentarily

I'll check that ASAP. Strange as nothing changed,,,

**Comment by nschlia** [comment]
After fiddling a lot with WinFSP, I finally managed to map a directory successfully. See https://github.com/nschlia/ffmpegfs/blob/windows/WINDOWS.md#usage

Basically, the drive syntax is important. Do not use the Cygwin-style declarations like /cydrive/c/xyz; use this: c:/xyz. The paths must be Windows-like; only replace all backslashes with forward slashes.

If you do not comply, WinFSP will retaliate with "Mount point in use" or by mapping the drive but not allowing anyone access to it, even the administrator. Or by throwing General Protection Faults. Very naughty.

Try this:

`ffmpegfs --audiobitrate=256K --videobitrate=1.5M   d:/tmp/in/ /cygdrive/D/tmp/out -o allow_other,ro,desttype=mp4`

But please note, as you are using Windows, the performance will be a tear-jerker. I waited several minutes for two 35 MB FLAC files to be available as MP3. If you seriously intend to use FFmpegfs, consider using a Linux box. Even a cheapy Raspberry 4 will do better than a 3.5 GHz Intel Machine under Windows 11.

Edit:

Update the sources and build first.

**Reply by kpcenti** [reply]
I wasn't sure how to update the source. So i just redownloaded the branch and rebuilt everything. I tried the new syntax and i'm still getting the same error
![image](https://github.com/user-attachments/assets/c6992f2a-2de0-411f-9ae5-403815193639)


I initially wanted to stick with Windows since I am familiar with it unlike Linux, and the software I wanted to use the output files with are on Windows. 

I strayed away from the Linux builds due to lack of knowledge how to set it up, and "Lock ups when accessed through Samba" "Songs get cut short" mentioned in the Problems doc thinking these problems won't occur in Windows

**Reply by nschlia** [reply]
Maybe Windows will work for you, but it does not perform good enough for me.

You could also try something like this:

`src/ffmpegfs --log_maxlevel=TRACE /cygdrive/D/tmp/in/ x: --desttype=mp3 --log_stderr`

Seems that mounting into a sub directory does not work for you, so maybe to a drive letter does.

**Reply by kpcenti** [reply]
It seems to be a file location error. I tried that exact command and it started to proceed. I changed the IN folder to a /cygdrive/c/TESTMUSICIN and kept everything else and it made a new virtual drive on x: (Super cool! I didn't know it could do this!)

Every enough trial and error with the location syntaxes I did manage to make it run but it claims to have no such file or directory:
![image](https://github.com/user-attachments/assets/f9f2e11e-5db1-4c29-9a91-7aafde2b6aa0)
It made a Local Disk (Y:) and has the first folder but nothing inside of it

The music i have is located on a Network Drive(since I'm using a VM) mapped to Z: in a folder called Navidrome Library, which has subfolders of music. I would like to mount it to either C:/ffmpegfs-output or a Virtual Drive like Y: since I could do that. Output being ALAC so I can have the music files appear in iTunes

Maybe WinFSP doesn't like network locations.

Not sure how I could proceed.

**Reply by kpcenti** [reply]
I played around with it some more and used an IN folder locally, and did W: and it seems to worked very slowly even with the error. Which is odd since it seems to not be running properly without all the text that appears when it gets mounted.

![image](https://github.com/user-attachments/assets/e80f5c24-14fd-4553-b341-0dab5dcef707)

EDIT:
I ran it with the logging and got this, but all the folders are blank 
![image](https://github.com/user-attachments/assets/c07abe09-e3bd-4bee-99cb-2c677f691c6e)

**Comment by kpcenti** [comment]
I think I will mark this as basically close. I mostly got it working just not in the directories I wanted, and at a snails pace when transcoding. I might submit some edits on the windows building directions to make it the most clear.


I've managed to get it installed in Linux. I didn't realize it was just one command in the terminal. 
Though I'm experiencing other problems once I get that running. I'll make a new thread in the discussions or issues if i can't get them working.


Thank you so much for the help!

**Reply by nschlia** [reply]
Thanks a lot for your help, too. Sorry it does not work as desired. I've also seen it acting strange sometimes, and, most of all, being slow.

I am sorry but I am unable to fix that on my side as this lies inside the WinFSP code. Maybe, if I find another API tool, I could get it working. But I have not seen anything so far.

The best solution would be to do the Fuse stuff myself, but apart from the time I'd have to spend it would also cost some money as it requires to buy certificates for the Windows drivers. 

It would be grant if you could clarify things in the building directions.
